rpc/handlers.go

```diff
@@ -775,9 +779,80 @@ func getAddInvokeTxCode(err error) int {
 		return jsonrpc.InvalidParams
 	case strings.Contains(err.Error(), "not supported. Supported versions"):
 		return jsonrpc.InvalidParams
-	case strings.Contains(err.Error(), "max_fee must be bigger than 0.\n0 >= "):
+	case strings.Contains(err.Error(), "max_fee must be bigger than 0"):
 		return jsonrpc.InvalidParams
 	default:
 		return jsonrpc.InternalError
 	}
 }
+
+// AddDeclareTransaction relays a declare transaction to the gateway.
+//
+// It follows the specification defined here:
+// https://github.com/starkware-libs/starknet-specs/blob/a789ccc3432c57777beceaa53a34a7ae2f25fda0/api/starknet_write_api.json#L39
+// Note: The gateway expects the sierra_program to be gzip compressed and 64-base encoded. We perform this operation,
+// and then relay the transaction to the gateway.
+func (h *Handler) AddDeclareTransaction(declareTx *json.RawMessage) (*DeclareTxResponse, *jsonrpc.Error) {
+	var request map[string]interface{}
+	err := json.Unmarshal([]byte(*declareTx), &request)
+	if err != nil {
+		return nil, jsonrpc.Err(jsonrpc.InternalError, err.Error())
+	}
+
+	if version, ok := request["version"]; ok && version == "0x2" {
+		contractClass, ok := request["contract_class"].(map[string]interface{})
+		if !ok {
+			return nil, jsonrpc.Err(jsonrpc.InvalidParams, errors.New("{'contract_class': ['Missing data for required field.']}").Error())
+		}
+		sierraProg, ok := contractClass["sierra_program"]
+		if !ok {
+			return nil, jsonrpc.Err(jsonrpc.InvalidParams, errors.New("{'sierra_program': ['Missing data for required field.']}").Error())
+		}
+
+		sierraProgBytes, errIn := json.Marshal(sierraProg)
+		if errIn != nil {
+			return nil, jsonrpc.Err(jsonrpc.InternalError, errIn.Error())
+		}
+
+		var gzipBuffer bytes.Buffer
+		gzipWriter := gzip.NewWriter(&gzipBuffer)
+		_, errIn = gzipWriter.Write(sierraProgBytes)
+		if errIn != nil {
+			return nil, jsonrpc.Err(jsonrpc.InternalError, errIn.Error())
+		}
+		if errIn = gzipWriter.Close(); errIn != nil {
+			return nil, jsonrpc.Err(jsonrpc.InternalError, errIn.Error())
+		}
+
+		gwSierraProg := base64.StdEncoding.EncodeToString(gzipBuffer.Bytes())
+		if errIn != nil {
+			return nil, jsonrpc.Err(jsonrpc.InternalError, errIn.Error())
+		}
```

Comment:
We do this in https://github.com/NethermindEth/juno/blob/main/starknetdata/feeder/feeder.go#L349-L360 as well. Let's extract that into a function and reuse that

Reply:
I've created a Gzip64Encode() function and placed it in utils.